crates/stages/stages/src/stages/era.rs

```diff
@@ -0,0 +1,603 @@
+                        let file = reth_fs_util::open(meta.path())?;
+                        let reader = Era1Reader::new(file);
+                        let iter = reader.iter();
+                        let iter = iter.map(era::decode);
+
+                        Ok(Box::new(iter) as Item<BH, BB>)
+                    })
+                })))
+            }
+        })
+    }
+
+    fn cleanup(&self) {
+        if let Self::Url(_, path) = self {
+            if let Ok(dir) = path.read_dir() {
+                for entry in dir.flatten() {
+                    let path = entry.path();
+
+                    if path.is_file() {
+                        let _ = reth_fs_util::remove_file(path);
+                    }
+                }
+            }
+        }
+    }
+}
+
+impl<BH: Debug, BB: Debug, F: Debug> Debug for EraStage<BH, BB, F> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("EraStage")
+            .field("source", &self.source)
+            .field("hash_collector", &self.hash_collector)
+            .field("item", &self.item.is_some())
+            .field("stream", &"dyn Stream")
+            .finish()
+    }
+}
+
+impl<BH, BB, F> EraStage<BH, BB, F> {
+    /// Creates a new [`EraStage`].
+    pub fn new(source: Option<F>, etl_config: EtlConfig) -> Self {
+        Self {
+            source,
+            item: None,
+            stream: None,
+            last_block_height: None,
+            hash_collector: Collector::new(etl_config.file_size, etl_config.dir),
+        }
+    }
+}
+
+impl<Provider, F, N> Stage<Provider> for EraStage<N::BlockHeader, N::BlockBody, F>
+where
+    Provider: DBProvider<Tx: DbTxMut>
+        + StaticFileProviderFactory<Primitives = N>
+        + BlockWriter<Block = N::Block>
+        + BlockReader<Block = N::Block>,
+    F: EraStreamFactory<N::BlockHeader, N::BlockBody> + Send + Sync + Clone,
+    N: NodePrimitives<BlockHeader: Value>,
+{
+    fn id(&self) -> StageId {
+        StageId::Era
+    }
+
+    fn poll_execute_ready(
+        &mut self,
+        cx: &mut Context<'_>,
+        input: ExecInput,
+    ) -> Poll<Result<(), StageError>> {
+        if input.target_reached() || self.item.is_some() {
+            return Poll::Ready(Ok(()));
+        }
+
+        if self.stream.is_none() {
+            if let Some(source) = self.source.clone() {
+                self.stream.replace(source.create(input)?);
+            }
+        }
+        if let Some(stream) = &mut self.stream {
+            if let Some(next) = ready!(stream.poll_next_unpin(cx))
+                .transpose()
+                .map_err(|e| StageError::Fatal(e.into()))?
+            {
+                self.item.replace(next);
+            }
+        }
+
+        Poll::Ready(Ok(()))
+    }
+
+    fn execute(&mut self, provider: &Provider, input: ExecInput) -> Result<ExecOutput, StageError> {
+        let era = self.item.take().ok_or(StageError::MissingDownloadBuffer)?;
```

Comment:
hmm, how does this behave once the entire era import finished, wouldn't this then always fail?

Reply:
Addressed in 9e83318be